scvi/external/sysvi/_model.py

```diff
@@ -0,0 +1,336 @@
+
+from scvi import REGISTRY_KEYS
+from scvi.data import AnnDataManager
+from scvi.data.fields import (
+    LayerField,
+    ObsmField,
+)
+from scvi.model.base import BaseModelClass, UnsupervisedTrainingMixin
+from scvi.utils import setup_anndata_dsp
+
+from ._module import SysVAE
+from ._trainingplans import TrainingPlanCustom
+from ._utils import prepare_metadata
+
+logger = logging.getLogger(__name__)
+
+
+class TrainingCustom(UnsupervisedTrainingMixin):
+    """Train method with custom TrainingPlan."""
+
+    # TODO could make custom Trainer (in a custom TrainRunner) to have in init params for early stopping
+    #  "loss" rather than "elbo" components in available param specifications - for now just use
+    #  a loss that is against the param specification
+
+    _training_plan_cls = TrainingPlanCustom
+
+
+class SysVI(TrainingCustom, BaseModelClass):
+    """Integration model based on cVAE with optional VampPrior and latent cycle-consistency loss.
+
+    Parameters
+    ----------
+    adata
+        AnnData object that has been registered via :meth:`~scvi-tools.SysVI.setup_anndata`.
+    prior
+        The prior distribution to be used. You can choose between "standard_normal" and "vamp".
+    n_prior_components
+        Number of prior components in VampPrior.
+    pseudoinputs_data_indices
+        By default VampPrior pseudoinputs are randomly selected from data.
+        Alternatively, one can specify pseudoinput indices using this parameter.
+    **model_kwargs
+        Keyword args for :class:`~scvi.external.sysvi.module.Module`
+    """
+
+    def __init__(
+        self,
+        adata: AnnData,
+        prior: Literal["standard_normal", "vamp"] = "vamp",
+        n_prior_components=5,
+        pseudoinputs_data_indices: np.array | None = None,
+        **model_kwargs,
+    ):
+        super().__init__(adata)
+
+        if prior == "vamp":
+            if pseudoinputs_data_indices is None:
+                pseudoinputs_data_indices = np.random.randint(
+                    0, adata.shape[0], n_prior_components
+                )
+            pseudoinput_data = next(
+                iter(
+                    self._make_data_loader(
+                        adata=adata,
+                        indices=pseudoinputs_data_indices,
+                        batch_size=n_prior_components,
+                        shuffle=False,
+                    )
+                )
+            )
+        else:
+            pseudoinput_data = None
+
+        n_cov_const = (
+            adata.obsm["covariates"].shape[1] if "covariates" in adata.obsm else 0
+        )
+        cov_embed_sizes = (
+            pd.DataFrame(adata.obsm["covariates_embed"]).nunique(axis=0).to_list()
+            if "covariates_embed" in adata.obsm
+            else []
+        )
+
+        # self.summary_stats provides information about anndata dimensions and other tensor info
+        self.module = SysVAE(
+            n_input=adata.shape[1],
```

Comment:
Preferable to access this via the `AnnDataManager` API

Reply:
See our other models.